File: tmp/20211127/273. 整数转换英文表示.ts
const radix = [1000000000, 1000000, 1000, 1]
const LESS_THAN_20 = [
  '',
  'One',
  'Two',
  'Three',
  'Four',
  'Five',
  'Six',
  'Seven',
  'Eight',
  'Nine',
  'Ten',
  'Eleven',
  'Twelve',
  'Thirteen',
  'Fourteen',
  'Fifteen',
  'Sixteen',
  'Seventeen',
  'Eighteen',
  'Nineteen',
]

const TENS = [
  '',
  'Ten',
  'Twenty',
  'Thirty',
  'Forty',
  'Fifty',
  'Sixty',
  'Seventy',
  'Eighty',
  'Ninety',
]

const THOUSANDS = ['Billion', 'Million', 'Thousand', '']

function numberToWords(num: number): string {
  if (num === 0) return 'Zero'
  const sb: string[] = []

  for (let i = 0; i < radix.length; i++) {
    const [div, mod] = [~~(num / radix[i]), num % radix[i]]
    if (div === 0) continue
    sb.push(trans(div))
    sb.push(THOUSANDS[i])
    sb.push(' ')
    num = mod
  }

  return sb.join('').trim()

  // 处理1000以下的数
  function trans(num: number): string {
    if (num === 0) return ''
    if (num < 20) return LESS_THAN_20[num] + ' '
    if (num < 100) return TENS[~~(num / 10)] + ' ' + trans(num % 10)
    return LESS_THAN_20[~~(num / 100)] + ' Hundred ' + trans(num % 100)
  }
}

console.log(numberToWords(1234567891))
// 输入：num = 1234567891
// 输出："One Billion Two Hundred Thirty Four Million Five Hundred Sixty Seven Thousand Eight Hundred Ninety One"

File: 0_字符串/字符串变换/273. 整数转换英文表示.py
# 273. 整数转换英文表示
# https://leetcode.cn/problems/integer-to-english-words/solution/zheng-shu-zhuan-huan-ying-wen-biao-shi-by-powcai/


from typing import List


class Solution:
    def numberToWords(self, num: int) -> str:
        to19 = (
            "One Two Three Four Five Six Seven Eight Nine Ten Eleven Twelve "
            "Thirteen Fourteen Fifteen Sixteen Seventeen Eighteen Nineteen".split()
        )
        tens = "Twenty Thirty Forty Fifty Sixty Seventy Eighty Ninety".split()

        def trans(num: int) -> List[str]:
            if num < 20:
                return to19[num - 1 : num]
            if num < 100:
                return [tens[num // 10 - 2]] + trans(num % 10)
            if num < 1000:
                return [to19[num // 100 - 1]] + ["Hundred"] + trans(num % 100)
            for p, w in enumerate(["Thousand", "Million", "Billion"], 1):
                if num < 1000 ** (p + 1):
                    return trans(num // 1000**p) + [w] + trans(num % 1000**p)
            return []

        return " ".join(trans(num)) or "Zero"
